Add support for swaps

danielabrozzoni opened the pull request:
Needs https://github.com/RGB-Tools/rust-lightning/pull/1/

danielabrozzoni committed:
Add vanilla channels and support for swaps

Co-authored-by: Alekos Filini <alekos.filini@gmail.com>

zoedberg commented on src/cli.rs:
```diff
@@ -1018,6 +1096,182 @@ pub(crate) async fn poll_for_user_input(
+							()
+						}),
+					};
+					if price.is_err() {
+						continue;
+					}
```
The command should not exit without telling to the user what is the encountered issue

zoedberg commented on src/cli.rs:
```diff
@@ -834,6 +861,46 @@ pub(crate) async fn poll_for_user_input(
+						PathBuf::from(&ldk_data_dir),
+					);
+				}
+				"coloredkeysend" => {
```
For uniformity with other new commands, I think that `coloredkeysend` and `keysend` should be implemented in the same match branch

zoedberg commented on src/cli.rs:
```diff
@@ -661,20 +667,38 @@ pub(crate) async fn poll_for_user_input(
 					println!("Refresh complete");
 				}
-				"openchannel" => {
+				"openchannel" | "opencoloredchannel" => {
```
I've noticed that the check for the minimum push amount (`if push_amt_msat < DUST_LIMIT_MSAT`) hasn't been disabled in case of vanilla channels, but IMO this check is not needed in that case

zoedberg commented on src/cli.rs:
```diff
@@ -1381,6 +1649,179 @@ fn list_payments(inbound_payments: PaymentInfoStorage, outbound_payments: Paymen
+	payment_storage: PaymentInfoStorage, ldk_data_dir: PathBuf,
+) {
+	let payment_preimage =
+		channel_manager.get_payment_preimage(payment_hash, payment_secret).unwrap();
```
Unwrapping this can cause a node to panic if a `makerexecute` command is provided with bad parameters

zoedberg commented on src/cli.rs:
```diff
@@ -1381,6 +1649,179 @@ fn list_payments(inbound_payments: PaymentInfoStorage, outbound_payments: Paymen
+		channel_manager.get_payment_preimage(payment_hash, payment_secret).unwrap();
+	// The cli takes the swaptype from the point of view of the user (=who wants to do the trade)
+	// In this function we're the market maker (=who sends the payment, completing the user trade)
+	// We swap the swaptype to opefully make this function easier to read
```
Typo: opefully -> hopefully

zoedberg commented on src/cli.rs:
```diff
@@ -1381,6 +1649,179 @@ fn list_payments(inbound_payments: PaymentInfoStorage, outbound_payments: Paymen
+	};
+
+	// Set swap flag
+	second_leg.paths[0].hops[0].short_channel_id |= 0x80000000_00000000;
```
Since this number (`0x80000000_00000000`) is used both here than in rust-lightning, I think it's better to define it as a costant in rust-lightning and import it here

zoedberg commented on src/main.rs:
```diff
@@ -177,62 +196,85 @@ async fn handle_ldk_events(
+			};
 
-			let rgb_utxos_path = format!("{}/rgb_utxos", ldk_data_dir);
+			let rgb_utxos_path = format!("{}/rgb_utxos", ldk_data_dir.display());
```
I don't think the call to `display` is needed here (and also in other points where you made this change) but maybe I'm missing something

danielabrozzoni replied:
We can either use `{}` and call `.display()` or use `{:?}`, they are equivalent

zoedberg commented on src/cli.rs:
```diff
@@ -1381,6 +1649,179 @@ fn list_payments(inbound_payments: PaymentInfoStorage, outbound_payments: Paymen
+		.await?;
+
+	let last_price = body.last_price.parse::<f64>()?;
+	let price = (1.0 / last_price * 1e11) as u64;
```
We should check that `last_price` is greater than `0`, otherwise we risk to have a division by zero that will panic the node

zoedberg reviewed (changes requested):
All things we individually reported and some linting (not reported), are small issues that we can fix in a commit post PR merge, if you prefer.

Something we spotted that needs to be addressed is improved test coverage. Specifically:
- there's no test for the "buy" swap type
- there's no test for the "rejecting non-whitelisted swap" case
- there's no test for routed swaps (i.e. more than a channel between taker/maker)

We've also noticed that when the swap type is "buy" you edit only the first path of the `first_leg` route by setting the fee to the `HTLC_MIN_MSAT` constant, so wondering if this is correct in case the route has multiple paths (i.e. multi-part payment).

Additionally, the `makerinit` command doesn't check if `price_msats_per_asset`, `expiry_secs` and `amount` are greater than `0`. Definitely `expiry_secs` needs to be positive (unless `0` is given the meaning of "it never expires"), probably it makes sense for the price to be `0` because it allows donations, but we'd like to know what's your opinion on this because maybe we're missing the reason these fields are allowed to be `0`. Also, when selling, the command doesn't check if the given `amount` is available (currently it's possible to sell more asset than the node balance or even more that the total issued supply). These checks should be eventually added also on the taker side, when it parses the swap string.

Finally, the `takerlist` command shows the whitelisted trades for the taker. We're wondering though how the maker should recover the list of initiated swaps (buy or sell) on his side.

zoedberg commented:
Sorry to report this only now but I've just noticed it. The `Event::SpendableOutputs` has not been changed, meaning that vanilla channels will fail to close (panicking the node). Please fix this and also improve vanilla tests by closing the channel and calling the `listunspent` command to check if the closing output appears.

danielabrozzoni commented:
> Something we spotted that needs to be addressed is improved test coverage. Specifically:
there's no test for the "buy" swap type
there's no test for the "rejecting non-whitelisted swap" case
there's no test for routed swaps (i.e. more than a channel between taker/maker)

Added tests for routed swaps and the buy swap type; the "rejecting non-whitelisted swap" case was present already (swap_roundtrip_fail)

> Additionally, the makerinit command doesn't check if price_msats_per_asset, expiry_secs and amount are greater than 0.

Added checks in makerinit and taker

> Also, when selling, the command doesn't check if the given amount is available (currently it's possible to sell more asset than the node balance or even more that the total issued supply).

This is currently not implemented in any of the rgb commands (keysend/payinvoice), do you think we could wait and implement it when we have it in the other commands as well?

> Finally, the takerlist command shows the whitelisted trades for the taker. We're wondering though how the maker should recover the list of initiated swaps (buy or sell) on his side.

We modified the command to be called `tradeslist` and optionally take a taker/maker argument for filtering.

> Sorry to report this only now but I've just noticed it. The Event::SpendableOutputs has not been changed, meaning that vanilla channels will fail to close (panicking the node). Please fix this and also improve vanilla tests by closing the channel and calling the listunspent command to check if the closing output appears.

Fixed.

zoedberg commented:
> This is currently not implemented in any of the rgb commands (keysend/payinvoice), do you think we could wait and implement it when we have it in the other commands as well?

The openchannel command actually logs `ERROR: do not have enough RGB assets` in case there are not enough RGB assets to open the channel. If we didn't put that check on the keysend/payinvoice APIs is just because we didn't find the time to do that yet. Anyway, I think these checks can be considered part of the job.

Do you have any news about what Matt proposed in https://github.com/lightningdevkit/rust-lightning/issues/2479 ?
